app/models/docker_registry.rb

```diff
@@ -0,0 +1,64 @@
+# model that represents a docker registry
+class DockerRegistry < ActiveRecord::Base
+  belongs_to :certificate, polymorphic: true
+
+  validates :url, presence: true
+  validates :url, uniqueness: true
+  validates :url, url: { schemes: ["https", "http"] }
+  class << self
+    def apply(settings_params)
+      errors = []
+      settings_params["registries"].each do |registry|
+        errors << configure_registry(registry) if registry["docker_registry_url"].present?
+      end
+      cleanup(settings_params)
+      errors
+    end
+
+    private
+
+    # create or update DockerRegistry model
+    def configure_registry(registry)
+      url     = registry["docker_registry_url"]
+      cert    = registry["docker_registry_certificate"]
+      mirror  = registry["docker_registry_mirror"]
+
+      certifiable_id, errors = configure_certificate(cert)
+
+      find_or_initialize_by(url: url).tap do |r|
+        r.url = url
+        r.certifiable_id = certifiable_id
+        r.certifiable_type = certifiable_id || :docker_registry
+        r.mirror = mirror
+        errors << "Registry url #{url} doesn't match a docker registry pattern" unless r.save
+      end
+
+      errors
+    end
+
+    # create or update Certificate model
+    def configure_certificate(cert)
+      errors = []
+      return [nil, errors] if cert.blank?
+      certificate = Certificate.find_or_initialize_by(certificate: cert).tap do |c|
+        c.certificate = cert
+        errors << "Failed to validate certificate" unless c.save
+      end
+      [certificate.id, errors]
+    end
+
+    # remove old registries from the db if they were deleted in the UI
+    def cleanup(settings_params)
+      passed_registries = settings_params["registries"].collect do |r|
+        r["docker_registry_url"]
+      end
+      saved_registries = all.map(&:url)
+
+      (saved_registries - passed_registries).each do |registry|
+        r = find_by(url: registry)
```

Comment:
Use transaction here?

Reply:
Or just a one liner: `DockerRegistry.where(url: saved_registries - passed_registries).destroy_all`

We should add a `before_destroy` or `after_destroy` to `DockerRegistry`, if it's the latest one referencing to a `Certificate` it will automatically `destroy` the certificate too.

Reply:
If what I wrote in https://github.com/kubic-project/velum/pull/193/files#r160642129 is right, we won't need that because we could use `dependent: :destroy` with `has_one` and just call `r.delete`.

Reply:
No, a `Certificate` can be linked to `DockerRegistry` or other entities in the future. Imagine a case in which:

`DockerRegistry` with `id` 1 is linked to `Certificate` with `id` 1
`DockerRegistry` with `id` 2 is linked to `Certificate` with `id` 1

If you add `dependent: :destroy`, if `DockerRegistry` with `id` `1` is removed, it will destroy the `Certificate` with `id` `1` leaving `DockerRegistry` orphan.

Reply:
You are right @vitoravelino, the polymorphic association is the other way around.

Reply:
Here's a suggestion because right now we are not handling properly the associations:

`CertificateService` (new model): `belongs_to :certificate` and `belongs_to :service, polymorphic: true` (so it has `certificate_id`, `service_id`, `service_type`)
`Certificate`: `has_many :certificate_service`

This would allow us to link the same certificate to any `service` (for now only `DockerRegistry`), update certificates only once and update them on several services at the same time, and also destroy `CertificateService` when a `service` is destroyed.